modules/tinymce/src/plugins/link/main/ts/ui/Controls.ts

```diff
@@ -105,28 +105,25 @@ const setupContextToolbars = (editor: Editor) => {
           return Actions.toggleActiveState(editor)(buttonApi);
         },
         onAction: (formApi) => {
-          const anchor = Utils.getAnchorElement(editor);
           const value = formApi.getValue();
-          if (!anchor) {
-            const attachState = { href: value, attach: Fun.noop };
-            const onlyText = Utils.isOnlyTextSelected(editor);
-            const text: Optional<string> = onlyText ? Optional.some(Utils.getAnchorText(editor.selection, anchor)).filter((t) => t.length > 0).or(Optional.from(value)) : Optional.none();
-            Utils.link(editor, attachState, {
-              href: value,
-              text,
-              title: Optional.none(),
-              rel: Optional.none(),
-              target: Optional.none(),
-              class: Optional.none()
-            });
-            formApi.hide();
-          } else {
-            editor.undoManager.transact(() => {
-              editor.dom.setAttrib(anchor, 'href', value);
-              collapseSelectionToEnd(editor);
-              formApi.hide();
-            });
-          }
+
+          // if we're editing a link, don't change the text.
+          // if anything other than text is selected, don't change the text.
+          const anchor = Utils.getAnchorElement(editor);
+          const onlyText = Utils.isOnlyTextSelected(editor);
+          const text: Optional<string> = !anchor && onlyText ? Optional.some(Utils.getAnchorText(editor.selection, anchor)).filter((t) => t.length > 0).or(Optional.from(value)) : Optional.none();
```

Comment:
Can we break this up a little bit? It's a very complicated ternary expression

Reply:
This is probably a good use case for `Optionals.someIf`. eg:
```suggestion
          const text = Optionals.someIf(!anchor && onlyText, Utils.getAnchorText(editor.selection, anchor))
            .filter((t) => t.length > 0)
            .or(Optional.from(value));
```

Reply:
Will do. I tend to avoid changing more than I need to (this code is almost entirely the same as it was before) but I'm happy to make further changes.

Reply:
> Remember to give the PR a more meaningful title, as that's going to be the commit message when it's merged

I always forget this repo changed to squash merge.

Reply:
It's not quite as simple as `someIf` - the `.or()` only runs inside the `if` condition of the ternary. But I think I have a decent refactoring for clarity.

Reply:
Actually the above wouldn't be the same because of the `or` at the end :cry: